test/vectorized_solve_Diff.jl

```diff
@@ -0,0 +1,226 @@
+@inline Base.IndexStyle(::Type{<:Diff}) = IndexCartesian()
+
+@inline function Base.size(D::Diff{T,N,DIMS}) where {T,N,DIMS}
+    sz = size(D.a)
+    return tupledec(sz, DIMS)
+end
+
+@inline function Base.getindex(D::Diff{T,N,DIMS}, I::Vararg{Int, N}) where {T,N,DIMS}
+    @boundscheck checkbounds(D, I...)
+    a = D.a
+    dI = tupleinc(I, DIMS)
+    @inbounds r = a[dI...] - a[I...]
+    return r
+end
+
+"Increment number at index i of tuple"
+tupleinc(t::Tuple, i) = tupleadd(t, i, 1)
+"Decrement number at index i of tuple"
+tupledec(t::Tuple, i) = tupleadd(t, i, -1)
+
+"Add number to tuple at one index"
+function tupleadd end
+tupleadd(t::NTuple{1}, i, n) = (t[1]+n,)
+tupleadd(t::NTuple{2}, i, n) = i==1 ? (t[1]+n, t[2]) : (t[1], t[2]+n)
+function tupleadd(t::NTuple{3}, i, n)
+    if i==1
+        (t[1]+n, t[2], t[3])
+    elseif i==2
+        (t[1], t[2]+n, t[3])
+    else
+        (t[1], t[2], t[3]+n)
+    end
+end
+function tupleadd(t::NTuple{4}, i, n)
+    if i==1
+        (t[1]+n, t[2], t[3], t[4])
+    elseif i==2
+        (t[1], t[2]+n, t[3], t[4])
+    elseif i==3
+        (t[1], t[2], t[3]+n, t[4])
+    else
+        (t[1], t[2], t[3], t[4]+n)
+    end
+end
+
+@views function vectorized_solve_v2( m :: Mesh, f :: Fields )
+
+    time   = 0.0
+    dx, dy = m.dx, m.dy
+    nx, ny = m.nx, m.ny
+
+    dtT = tetT*1/4.1*min(dx,dy)^2 # explicit timestep for 2D diffusion
+
+    errs  = []
+
+    etav    = zeros(Float64,(nx+1,ny+1))
+    Exyv    = zeros(Float64,(nx+1,ny+1))
+    Exyc    = zeros(Float64,(nx,ny))
+    divV    = zeros(Float64,(nx,ny))
+    Exxc    = zeros(Float64,(nx,ny))
+    Eyyc    = zeros(Float64,(nx,ny))
+    Eii2    = zeros(Float64,(nx,ny))
+    etac_phys = zeros(Float64,(nx,ny))
+    Sxx     = zeros(Float64,(nx,ny))
+    Syy     = zeros(Float64,(nx,ny))
+    Txy     = similar(Exyv)
+    Hs      = zeros(Float64,(nx,ny))
+    To      = similar(f.T)
+    dPdtauP = similar(divV)
+    dTdtauT = similar(Hs)
+    dtauP   = similar(f.etac)
+    dtauT   = similar(f.etac)
+    dtauVx  = zeros(Float64,(nx-1,ny))
+    dtauVy  = zeros(Float64,(nx,ny-1))
+
+    for it = 1:nt # Physical timesteps
+
+        # temperature from previous step (for backward-Euler integration)
+        To .= f.T
+
+        time += dtT # update physical time
+
+        for iter = 1:niter # Pseudo-Transient cycles
+
+            # used for damping x momentum residuals
+            f.dVxdtauVx0 .= f.dVxdtauVx .+ dampx .* f.dVxdtauVx0
+
+            # used for damping y momentum residuals
+            f.dVydtauVy0 .= f.dVydtauVy .+ dampy .* f.dVydtauVy0
+
+            #  Kinematics
+            f.Vx[  :,  1] .= f.Vx[    :,    2]
+            f.Vx[  :,end] .= f.Vx[    :,end-1]
+            f.Vy[  1,  :] .= f.Vy[    2,    :]
+            f.Vy[end,  :] .= f.Vy[end-1,    :]
+
+            divV .= (Diff(f.Vx,1)[:,2:end-1] ./ dx
+                  .+ Diff(f.Vy,2)[2:end-1,:] ./ dy)
+
+            Exxc .= Diff(f.Vx,1)[:,2:end-1] ./ dx .- 1/2 .* divV
```

Comment:
Note that whilst `Diff(f.Vx,1)[:,2:end-1]==Diff(f.Vx[:,2:end-1],1)`, the latter leads to allocations (not sure why).